### PR TITLE
Add stream partitioning feature guide

### PR DESCRIPTION
**Pull request by ilayaperumalg**

### CODE REVIEW

File: content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md [hunk 1]
```diff
@@ -4,6 +4,138 @@ title: 'Data Partitioning'
 description: 'How to use partitioning'
 ---
 
-Partitioning support allows for content-based routing of payloads to downstream application instances in an event streaming pipeline.
-This is especially useful when you want to have your downstream application instances processing data from specific partitions.
+Partitioning is a critical concept in stateful processing, for either performance or consistency reasons, to ensure that all related data is processed together.
+For example, in a time-windowed average calculation example, it is important that all measurements from any given sensor are processed by the same application instance.
+Alternatively, you may want to cache some data related to the incoming events so that it can be enriched without making a remote procedure call to retrieve the related data.
+
+Partitioning support allows for `content` based routing of payloads to the `downstream` application instances in a streaming data pipeline.
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md]: It'd be good to remove backticks from content and downstream - they are not code fragments

File: content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md [hunk 2]
```diff
@@ -4,6 +4,138 @@ title: 'Data Partitioning'
...
+
+- **app.[app/label name].producer.partitionSelectorClass** - The class name of a PartitionSelectorStrategy (default: null)
+
+- **app.[app/label name].producer.partitionSelectorExpression** - A SpEL expression, evaluated against the partition key, to determine the partition index to which the message is routed. The final partition index is the return value (an integer) modulo [nextModule].count. If both the class and expression are null, the underlying binder’s default PartitionSelectorStrategy is applied to the key (default: null)
+
+In summary, an app is partitioned if its count is > 1 and the previous app has a partitionKeyExtractorClass or partitionKeyExpression (partitionKeyExtractorClass takes precedence). When a partition key is extracted, the partitioned app instance is determined by invoking the partitionSelectorClass, if present, or the partitionSelectorExpression % partitionCount. partitionCount is application count, in the case of RabbitMQ, or the underlying partition count of the topic, in the case of Kafka.
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md]: > .. if its deployment instances count is >1 ...

File: content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md [hunk 3]
```diff
@@ -4,6 +4,138 @@ title: 'Data Partitioning'
...
+
+#### Create the stream
+
+In this section, you will see how a partitioned stream can be created and deployed.
+
+Let's consider the following stream:
+
+- `http` source application listens at port 9001 for incoming text/sentence
+- `splitter` processor application splits the sentence into words and partition the words based on their hash value (by using the `payload` as the partitionKeyExpression)
+- `log` sink application is scaled to run `three` application instances and each instance is expected to receive unique hash values from the upstream
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/3-partitioning.md]: remove backticks for three

**Comment by markpollack** [comment]
squashed and merged as f072b5ed334f9cb4ff0d427c62fd2bd93d43632d

**Pull request closed without merging**